File: CS50X Project/Week 06/sentimental-credit/credit.py
import re
import math

def main():
    verify_card(input("Number: "))


def verify_card(num):

    if verify_digits(int(num)):

        if re.search(r"^3(4|7)\d{13}$", num):
            print("AMEX")

        elif re.search(r"^5[1-5]\d{14}$", num):
            print("MASTERCARD")

        elif re.search(r"^4\d{12}(?:\d{3})?$", num):
            print("VISA")

        else:
            print("INVALID")

    else:
        print("INVALID")


def verify_digits(num):

    sum = sumodd(num, 0) + sumeven(math.floor(num/10), 0)
    return sum % 10 == 0

def sumodd(n, sum):
    if n > 0:
        return sumodd(math.floor(n/100), sum + n % 10)
    else:
        return sum

def sumeven(n, sum):
    if n > 0:
        helper = (n % 10) * 2
        if helper >= 10:
            helper = helper % 10 + 1
        return sumeven(math.floor(n/100), sum + helper)
    else:
        return sum


if __name__ == "__main__":
    main()

File: CS50X Project/Week 01/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int sum2posi(long n, int summing);
int sum1posi(long n, int summing);

int main(void)
{
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 0);

    int sum = sum2posi((num - num % 10) / 10, 0) + sum1posi(num, 0);
    if (sum % 10 == 0)
    {
        if ((4000000000000 <= num && num < 5000000000000) ||
            (4000000000000000 <= num && num < 5000000000000000))
        {
            printf("VISA\n");
        }
        else if (5100000000000000 <= num && num < 5600000000000000)
        {
            printf("MASTERCARD\n");
        }
        else if ((340000000000000 <= num && num < 350000000000000) ||
                 (370000000000000 <= num && num < 380000000000000))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int sum2posi(long n, int summing)
{
    if (n > 0)
    {
        int sum_helper = (n % 10) * 2;
        if (sum_helper >= 10)
        {
            sum_helper = sum_helper % 10 + 1;
        }
        return sum2posi((n - n % 100) / 100, summing + sum_helper);
    }
    else
    {
        return summing;
    }
}

int sum1posi(long n, int summing)
{
    if (n > 0)
    {
        return sum1posi((n - n % 100) / 100, summing + (n % 10));
    }
    else
    {
        return summing;
    }
}
